### PR TITLE
ENH: add support for filtered loading with open_file using mask and rows

### LINKED ISSUE #1157: Lazy Loading Dataframes
There have been times when I wanted to use geopandas to get an idea of what was in a large file. But, it took a long time to load it in. So, I used `fiona` instead. Are there any plans (or desires) to do lazy loading of the geopandas data?

This is library I recently saw is related:
https://vaex.readthedocs.io/en/latest/index.html
https://github.com/vaexio/vaex

### PR DESCRIPTION
**Pull request by snowman2**
Closes #1157 

Opening PR so you can see the code and recommend changes. If you like the implementation, I can add some tests.

### CODE REVIEW

**Comment by martinfleis** [comment]
I like this!

`slice` looks perfect. In terms of `mask`, which expects geometry (correct?), what can we pass? GeoDataFrame, GeoSeries, Polygon?? Can we support all of them?

**Comment by snowman2** [comment]
Sounds good 👍

I wasn't sure that to do about the geoseries for the mask.  Maybe just do a union of the geometries?

**Comment by martinfleis** [comment]
> Maybe just do a union of the geometries?

Yep. `geometry.unary_union` will do.

**Comment by snowman2** [comment]
Sounds like a good plan to me. I will update it with that change later.

**Comment by snowman2** [comment]
Hmmm:

```
RemoveError: 'setuptools' is a dependency of conda and cannot be removed from

conda's operating environment.

The command "conda update conda" failed and exited with 1 during .
```

**Comment by martinfleis** [comment]
@snowman2 known issue https://github.com/conda/conda/issues/9337

**Comment by jorisvandenbossche** [comment]
Some questions:

- How does the mask work? Is it intersects / contains / ... based? Is it using the actual shape or just the bounding box?
- Can you show a few benchmarks? Eg check the timing of reading the full file vs masking or slicing a small portion of it?

**Comment by snowman2** [comment]
> How does the mask work? 

The mask uses the setspatialfilter: https://gdal.org/doxygen/classOGRLayer.html#a0b4ab45cf97cbc470f0d60474d3e4169
```
This method set the geometry to be used as a spatial filter when fetching features via the GetNextFeature() method. Only features that geometrically intersect the filter geometry will be returned.

Currently this test is may be inaccurately implemented, but it is guaranteed that all features whose envelope (as returned by OGRGeometry::getEnvelope()) overlaps the envelope of the spatial filter will be returned. This can result in more shapes being returned that should strictly be the case.

Starting with GDAL 2.3, features with null or empty geometries will never be considered as matching a spatial filter.

This method makes an internal copy of the passed geometry. The passed geometry remains the responsibility of the caller, and may be safely destroyed.

For the time being the passed filter geometry should be in the same SRS as the layer (as returned by OGRLayer::GetSpatialRef()). In the future this may be generalized.
```

**Comment by snowman2** [comment]
Slicing benchmark:
https://github.com/geopandas/geopandas/issues/1157#issuecomment-541480982
image

File: geopandas/io/file.py [hunk 1]
```diff
@@ -83,11 +94,31 @@ def read_file(filename, bbox=None, **kwargs):
...
-                f_filt = features.filter(bbox=bbox)
+            # handle loading the mask
+            elif isinstance(mask, (GeoDataFrame, GeoSeries)):
+                mask = mapping(mask.to_crs(crs).unary_union)
+            elif isinstance(mask, BaseGeometry):
+                mask = mapping(mask)
+            # setup the data loading filter
+            if data_slice is not None:
+                if not isinstance(data_slice, slice):
+                    raise TypeError("'data_slice' must be a slice.")
```
Comment (martinfleis) [inline geopandas/io/file.py]: Can you test this error as well? Somtething like this:

```py
with pytest.raises(TypeError):
    read_file(nybb_filename, data_slice="not_a_slice")
```

Comment (snowman2) [reply geopandas/io/file.py]: Oh yeah, good catch.

File: doc/source/io.rst [hunk 2]
```diff
@@ -31,10 +31,69 @@ a web URL, for example for GeoJSON files from `geojson.xyz <http://geojson.xyz/>
...
+
+.. versionadded:: 0.7.0
+
+The geometry filter only loads data that intersects with the geometry.
+
+.. code-block:: python
+
+    from shapely.geometry import box
+
+    mask = box(
```
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: I guess that does confuse things with the bbox. I like your idea.

[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: How is the current iteration?

[the remark this answers is not in the document]
Comment (martinfleis) [reply doc/source/io.rst]: Perfect!

[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Great, thanks for the idea!

File: geopandas/io/file.py [hunk 3]
```diff
@@ -38,18 +40,27 @@ def _is_url(url):
...
+    bbox: tuple | GeoDataFrame or GeoSeries | shapely Geometry, default None
+        Filter features by given bounding box, GeoSeries, GeoDataFrame or a
+        shapely geometry. CRS mis-matches are resolved if given a GeoSeries
+        or GeoDataFrame. Cannot be used with mask.
+    mask: dict | GeoDataFrame or GeoSeries | shapely Geometry, default None
+        Filter features by given dict-like geojson geometry, GeoSeries,
+        GeoDataFrame or a shapely geometry. CRS mis-matches are resolved
+        if given a GeoSeries or GeoDataFrame. Cannot be used with bbox.
+    data_slice: slice, default None
+        Load in specific rows by passing in a slice() object.
```
Comment (jorisvandenbossche) [inline geopandas/io/file.py]: Some random thoughts here: 

- doing `data_slice=slice(...)` gives a duplicate "slice" usage. Wondering if we can have another keyword name that avoids this
- I think that it would be nice to support the basic case of reading the first n rows as well (expanding to `slice(n) == slice(None, n)`).

File: doc/source/io.rst [hunk 4]
```diff
@@ -31,10 +31,67 @@ a web URL, for example for GeoJSON files from `geojson.xyz <http://geojson.xyz/>
...
 
 
+Reading subsets of the data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since geopandas is powered by `Fiona`, which is powered by `GDAL`, you can take advantage of
+pre-filtering when loading in larger datasets. This can be done geospatially with a geometry
+or with a slice. Read more at :func:`geopandas.read_file`.
+
+Geometry Filter:
```
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Sure.

File: doc/source/io.rst [hunk 5]
```diff
@@ -31,10 +31,67 @@ a web URL, for example for GeoJSON files from `geojson.xyz <http://geojson.xyz/>
...
+        geopandas.datasets.get_path("naturalearth_cities"),
+        mask=gdf_mask[gdf_mask.continent=="Africa"],
+    )
+
+Data Slice Filter:
+^^^^^^^^^^^^^^^^^^
+
+.. versionadded:: 0.7.0
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Sure.

File: doc/source/io.rst [hunk 6]
```diff
@@ -31,10 +31,67 @@ a web URL, for example for GeoJSON files from `geojson.xyz <http://geojson.xyz/>
...
+        bbox=mask,
+    )
+
+
 Writing Spatial Data
 ---------------------
 
-GeoDataFrames can be exported to many different standard formats using the ``GeoDataFrame.to_file()`` method. For a full list of supported formats, type ``import fiona; fiona.supported_drivers``.
+GeoDataFrames can be exported to many different standard formats using the
+:meth:`~geopandas.GeoDataFrame.to_file()` method. For a full list of supported
```
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Agreed.

File: geopandas/io/file.py (hunk 3 above)
Comment (snowman2) [reply geopandas/io/file.py]: > Wondering if we can have another keyword name that avoids this

Have anything in particular in mind? `row_filter`?

File: doc/source/io.rst [hunk 7]
```diff
@@ -31,10 +31,67 @@ a web URL, for example for GeoJSON files from `geojson.xyz <http://geojson.xyz/>
...
+.. code-block:: python
+
+    gdf = geopandas.read_file(
+        geopandas.datasets.get_path("naturalearth_lowres"),
+        data_slice=slice(0, 10),
+    )
+
+
+Bounding Box Filter:
+^^^^^^^^^^^^^^^^^^^^
```
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Sure.

File: doc/source/io.rst [hunk 8]
```diff
@@ -31,10 +31,67 @@ a web URL, for example for GeoJSON files from `geojson.xyz <http://geojson.xyz/>
 *geopandas* can also get data from a PostGIS database using the ``read_postgis()`` command.
 
 
+Reading subsets of the data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Since geopandas is powered by `Fiona`, which is powered by `GDAL`, you can take advantage of
+pre-filtering when loading in larger datasets. This can be done geospatially with a geometry
```
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Good thinking.

File: doc/source/io.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (snowman2) [reply doc/source/io.rst]: Hmm, that broke the link. I will use a middle ground approach.

File: geopandas/io/file.py (hunk 3 above)
Comment (snowman2) [reply geopandas/io/file.py]: I kinda like `row_filter` and am going with that one.

**Comment by martinfleis** [comment]
@snowman2 can you resolve merge conflicts?

**Comment by snowman2** [comment]
@martinfleis, just resolved the conflicts.

**Comment by snowman2** [comment]
Looks like everything related passes. Unrelated failures in the build:

```
- TestSetitem.test_setitem_slice[True] 
- test_setitem_loc_iloc_slice
- test_setitem_slice
```

File: doc/source/io.rst [hunk 9]
```diff
@@ -42,13 +42,72 @@ If there are multiple datasets in a folder in the ZIP file, you also have to spe
...
+The bounding box filter only loads data that intersects with the bounding box.
+
+.. code-block:: python
+
+    bbox = (
+        1031051.7879884212, 224272.49231459625, 1047224.3104931959, 244317.30894023244
+    )
+    gdf = geopandas.read_file(
+        geopandas.datasets.get_path("nybb"),
+        bbox=mask,
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]:
```suggestion
        bbox=bbox,
```

File: doc/source/io.rst [hunk 10]
```diff
@@ -42,13 +42,72 @@ If there are multiple datasets in a folder in the ZIP file, you also have to spe
...
+    )
+
 
 
 Writing Spatial Data
 ---------------------
 
-GeoDataFrames can be exported to many different standard formats using the ``GeoDataFrame.to_file()`` method. For a full list of supported formats, type ``import fiona; fiona.supported_drivers``.
+GeoDataFrames can be exported to many different standard formats using the
+:meth:`geopandas.GeoDataFrame.to_file()` method.
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]:
```suggestion
:meth:`geopandas.GeoDataFrame.to_file` method.
```

File: geopandas/io/file.py [hunk 11]
```diff
@@ -30,18 +32,27 @@ def _is_url(url):
...
-    bbox : tuple | GeoDataFrame or GeoSeries, default None
-        Filter features by given bounding box, GeoSeries, or GeoDataFrame.
-        CRS mis-matches are resolved if given a GeoSeries or GeoDataFrame.
+    bbox: tuple | GeoDataFrame or GeoSeries | shapely Geometry, default None
+        Filter features by given bounding box, GeoSeries, GeoDataFrame or a
+        shapely geometry. CRS mis-matches are resolved if given a GeoSeries
+        or GeoDataFrame. Cannot be used with mask.
+    mask: dict | GeoDataFrame or GeoSeries | shapely Geometry, default None
+        Filter features by given dict-like geojson geometry, GeoSeries,
+        GeoDataFrame or a shapely geometry. CRS mis-matches are resolved
```
Comment (jorisvandenbossche) [inline geopandas/io/file.py]: Can you clarify *how* features are filtered? (you answered on that to a previous question of me about that with a quote from gdal, I think it uses intersects) 
As otherwis it is not very clear what the difference is with bbox

**Comment by jorisvandenbossche** [comment]
Sorry to come back once more on the keyword names, but: the thing I don't really like about `row_filter` is that it is not a "filter" in general (at least not in the sense that I explain pandas indexing, where you would typically "filter" with a boolean mask based on a condition). 

Other proposal: we could also call it `nrows` and allow `nrows=n` (which would be expanded to a slice of `slice(0, n)` to read the first n rows). 
This makes the usage of "reading the first n rows" (which I *suppose* will be the most common case?) easier, and consistent with how that is done in pandas `read_` functions. And compared to pandas, the `nrows` keyword would in addition also support an actual slice object to not start from row 0 but read a number of rows somewhere in the middle. 
(this makes it of course somewhat inconsistent with the word "nrows", which is also not ideal though ..)

**Comment by snowman2** [comment]
How about `isel`? Or `isel_rows`? It would be analogous to pandas `.isel()`

File: geopandas/io/file.py (hunk 11 above)
Comment (snowman2) [reply geopandas/io/file.py]: Just updated. Thoughts?

**Comment by jorisvandenbossche** [comment]
I think `isel` is from xarray-terminology and not pandas?

File: geopandas/io/file.py (hunk 11 above)
Comment (jorisvandenbossche) [reply geopandas/io/file.py]: Looks good!

**Comment by martinfleis** [comment]
Isn't `isel` from xarray rather than pandas? What about simple `rows`?

**Comment by snowman2** [comment]
> I think isel is from xarray-terminology and not pandas?

Oh yeah. mixing that in my head.

**Comment by snowman2** [comment]
> What about simple rows?

I like that :+1:

**Comment by jorisvandenbossche** [comment]
`rows` is fine for me as well. I think `nrows` has a small plus in being consistent with pandas, `rows` on the other hand might capture the slice a bit better.

**Comment by martinfleis** [comment]
>  allow nrows=n (which would be expanded to a slice of slice(0, n) to read the first n rows)

I quite like this idea as well, I can imagine that is something I would personally do and it is easier to do `rows=10` than remembering to do `rows=slice(0, 10)`.

**Comment by snowman2** [comment]
> allow nrows=n (which would be expanded to a slice of slice(0, n) to read the first n rows)

Just added that in.

File: geopandas/io/tests/test_file.py [hunk 12]
```diff
@@ -1,24 +1,23 @@
...
-import fiona
 import numpy as np
+
+import fiona
+import geopandas
 import pandas as pd
+import pytest
 from shapely.geometry import Point, Polygon, box
 
-import geopandas
```
Comment (jorisvandenbossche) [inline geopandas/io/tests/test_file.py]: Just a question, do you have some automatic (isort?) hook set up that changes all those imports? If that is the case, maybe better turn it off in the geopandas code base as it gives quite some unrelated changes.

Comment (snowman2) [reply geopandas/io/tests/test_file.py]: Yeah, I do. I have it disabled now.

**Comment by jorisvandenbossche** [comment]
@snowman2 Thanks a lot !!

**Comment by snowman2** [comment]
@jorisvandenbossche  and @martinfleis, thanks for getting this in :+1:

File: geopandas/io/tests/test_file.py (hunk 12 above)
Comment (jorisvandenbossche) [reply geopandas/io/tests/test_file.py]: BTW, while I just asked to turn off automatic formatting .. ;) 
I would personally recommend using git pre-commit hook when working on geopandas (we have configuration for it included in the repo). That way you won't need to add commits with "blacken", as that gets done automatically. And it ensures it are the same checks as done online.

Comment (snowman2) [reply geopandas/io/tests/test_file.py]: I normally have all of that configured. But, it is different for each project. I didn't have one specific to geopandas setup and was trying to get this in quick - I figured it would be a small change 🙃. But, probably would have been worth the time to get setup.
